(is_non_overlapping_and_dense) gso to guard_or_false in when checking length 1

ColinPeppler opened the pull request:
Switch from `guard_size_oblivious` to `guard_or_false` if you encounter a DDE, this would then fallback to computing elementwise strides.

https://github.com/pytorch/pytorch/blob/2dccff7dcf56b0d168ebfd7ca08bdeca37273c56/torch/_prims/__init__.py#L1919-L1923

We think it's safe because Laith tested whether this fallback would fail any tests. It did not.
https://github.com/pytorch/pytorch/pull/158157

## Data-dependent exceptions (DDE)
```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 494, in is_non_overlapping_and_dense
    if guard_size_oblivious(length == 1):
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0 - 4, 1) (unhinted: Eq(u0 - 4, 1)).  (Size-like symbols: u0)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #159184
* __->__ #158894



cc @ezyang @penguinwu @bobrenjc93 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

ColinPeppler committed:
(is_non_overlapping_and_dense) gso to guard_or_false in when checking length-1 & expected stride

[ghstack-poisoned]

ColinPeppler committed:
Update on "(is_non_overlapping_and_dense) gso to guard_or_false in when checking length-1 & expected stride"

Switch from `guard_size_oblivious` to `guard_or_false` if you encounter a DDE, this would then fallback to computing elementwise strides.

https://github.com/pytorch/pytorch/blob/2dccff7dcf56b0d168ebfd7ca08bdeca37273c56/torch/_prims/__init__.py#L1919-L1923

We think it's safe because Laith tested whether this fallback would fail any tests. It did not.
https://github.com/pytorch/pytorch/pull/158157

## Data-dependent exceptions (DDE)
```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 494, in is_non_overlapping_and_dense
    if guard_size_oblivious(length == 1):
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0 - 4, 1) (unhinted: Eq(u0 - 4, 1)).  (Size-like symbols: u0)
```

```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 500, in is_non_overlapping_and_dense
    if guard_size_oblivious(stride != expected_stride):

Caused by: unbacked_size.int(),  # test/inductor/test_unbacked_symints.py:538 in fn (_prims_common/__init__.py:500 in is_non_overlapping_and_dense)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ColinPeppler committed:
Update on "(is_non_overlapping_and_dense) gso to guard_or_false in when checking length-1 & expected stride"

Switch from `guard_size_oblivious` to `guard_or_false` if you encounter a DDE, this would then fallback to computing elementwise strides.

https://github.com/pytorch/pytorch/blob/2dccff7dcf56b0d168ebfd7ca08bdeca37273c56/torch/_prims/__init__.py#L1919-L1923

We think it's safe because Laith tested whether this fallback would fail any tests. It did not.
https://github.com/pytorch/pytorch/pull/158157

## Data-dependent exceptions (DDE)
```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 494, in is_non_overlapping_and_dense
    if guard_size_oblivious(length == 1):
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0 - 4, 1) (unhinted: Eq(u0 - 4, 1)).  (Size-like symbols: u0)
```

```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 500, in is_non_overlapping_and_dense
    if guard_size_oblivious(stride != expected_stride):

Caused by: unbacked_size.int(),  # test/inductor/test_unbacked_symints.py:538 in fn (_prims_common/__init__.py:500 in is_non_overlapping_and_dense)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ColinPeppler committed:
Update on "(is_non_overlapping_and_dense) gso to guard_or_false in when checking length-1 & expected stride"

Switch from `guard_size_oblivious` to `guard_or_false` if you encounter a DDE, this would then fallback to computing elementwise strides.

https://github.com/pytorch/pytorch/blob/2dccff7dcf56b0d168ebfd7ca08bdeca37273c56/torch/_prims/__init__.py#L1919-L1923

We think it's safe because Laith tested whether this fallback would fail any tests. It did not.
https://github.com/pytorch/pytorch/pull/158157

## Data-dependent exceptions (DDE)
```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 494, in is_non_overlapping_and_dense
    if guard_size_oblivious(length == 1):
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0 - 4, 1) (unhinted: Eq(u0 - 4, 1)).  (Size-like symbols: u0)
```

```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 500, in is_non_overlapping_and_dense
    if guard_size_oblivious(stride != expected_stride):

Caused by: unbacked_size.int(),  # test/inductor/test_unbacked_symints.py:538 in fn (_prims_common/__init__.py:500 in is_non_overlapping_and_dense)
```




cc ezyang penguinwu bobrenjc93 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ezyang commented on torch/_prims_common/__init__.py:
```diff
@@ -491,10 +494,10 @@ def __eq__(self, other):
             continue
 
-        if guard_size_oblivious(stride != expected_stride):
+        if guard_or_false(stride != expected_stride):
```
NGL, this is pretty suspicious. Let's suppose I have a tensor with size `(u0, u1)` and stride `(u2, u3)`. If we somehow get to this loop, we will never trigger any of the conditions here, and then we will return True (that it is non overlapping and dense). But... it's clearly NOT??? Like I can trivially fill in values of u0..u3 that would make it overlapping. It is MUCH safer to guard_or_true here.

ColinPeppler replied:
That's a good point! I originally did not see it this way. This will be a better test case.

ColinPeppler committed:
Update on "(is_non_overlapping_and_dense) gso to guard_or_false in when checking length-1 & expected stride"

Switch from `guard_size_oblivious` to `guard_or_false` if you encounter a DDE, this would then fallback to computing elementwise strides.

https://github.com/pytorch/pytorch/blob/2dccff7dcf56b0d168ebfd7ca08bdeca37273c56/torch/_prims/__init__.py#L1919-L1923

We think it's safe because Laith tested whether this fallback would fail any tests. It did not.
https://github.com/pytorch/pytorch/pull/158157

## Data-dependent exceptions (DDE)
```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 494, in is_non_overlapping_and_dense
    if guard_size_oblivious(length == 1):
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0 - 4, 1) (unhinted: Eq(u0 - 4, 1)).  (Size-like symbols: u0)
```

```
  File "/data/users/colinpeppler/pytorch/torch/_decomp/decompositions.py", line 2139, in _to_copy
    x_tensor = torch._prims.convert_element_type(x_tensor, dtype)
  ...
  File "/data/users/colinpeppler/pytorch/torch/_prims/__init__.py", line 1920, in _convert_element_type_meta
    if torch._prims_common.is_non_overlapping_and_dense(a):
  File "/data/users/colinpeppler/pytorch/torch/_prims_common/__init__.py", line 500, in is_non_overlapping_and_dense
    if guard_size_oblivious(stride != expected_stride):

Caused by: unbacked_size.int(),  # test/inductor/test_unbacked_symints.py:538 in fn (_prims_common/__init__.py:500 in is_non_overlapping_and_dense)
```




cc ezyang penguinwu bobrenjc93 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

pianpwk reviewed (approved):
smaller change seems straightforward enough

ColinPeppler commented:
@pytorchbot merge

laithsakka commented:
I will approve to unblock current user, but I would probably rewrite this by introducing is_non_overlapping_and_dense_or_false and use it in the _convert_element_type_meta, that would be the unbacked semantics of that part.
we already have an is_non_overlapping_and_dense_or_false function in c++.

laithsakka commented on torch/_prims_common/__init__.py:
```diff
@@ -491,7 +494,7 @@ def __eq__(self, other):
+        if guard_or_false(length == 1):
             continue
 
         if guard_size_oblivious(stride != expected_stride):
```
you can do guard_or_true here and rename the function to is_non_overlapping_and_dense_or_false.

laithsakka replied:
its ok i  will address those in a different PR

laithsakka commented:
I think i have an idea how to get rid of the size oblivious here, the main idea is the following:
given unbacked strides thats the worse case that can happen if we do wrong sorting?

Answer : we would return False ! lol so not the end of the world if the function definitions is 
# may return false for unbacked input instead of throwing DDE.

def is_non_overlapping_and_dense_or_false(a: Tensor) -> bool:
that said we should try to do correct storting as much as we can using the Mod idea.
